compaction/task_manager_module.cc

```diff
@@ -330,25 +330,31 @@ tasks::is_abortable compaction_task_impl::is_abortable() const noexcept {
     return tasks::is_abortable{!_parent_id};
 }
 
-static future<bool> maybe_flush_all_tables(sharded<replica::database>& db) {
-    auto interval = db.local().get_config().compaction_flush_all_tables_before_major_seconds();
-    if (interval) {
-        auto when = db_clock::now() - interval * 1s;
-        if (co_await replica::database::get_all_tables_flushed_at(db) <= when) {
-            co_await db.invoke_on_all([&] (replica::database& db) -> future<> {
-                co_await db.flush_all_tables();
-            });
-            co_return true;
+static future<bool> maybe_flush_all_tables(sharded<replica::database>& db, flush_mode flush_mode, bool force_purge_tombstones) {
+    // flush all tables either if
+    // (a) force_purge_tombstone == true (or)
+    // (b) flush_mode == flush_mode::all_tables and compaction_flush_all_tables_before_major_seconds has elapsed since last all tables flush
+    if (!force_purge_tombstones) {
+        if (flush_mode != flush_mode::all_tables) {
+            co_return false;
+        }
+        auto interval = db.local().get_config().compaction_flush_all_tables_before_major_seconds();
```

Comment:
Hmm, the point of this timeout was to flush all memtables only once when major-compacting several (or all) keyspaces in sequence, so only the first major compaction would flush all memtables and the subsequent keyspace-major compactions would flush only their respective tables' memtables.

With force_purge_tombstones this patch means we'll flush all memtables for every keyspace.
It's not terrible.  I think we can live with it since it's an admin operation that is expected to be infrequent.